File: Java/Eta/com/thomsonreuters/upa/valueadd/common/VaQueue.java
package com.thomsonreuters.upa.valueadd.common;

/** A FIFO queue. */
public class VaQueue
{
    VaNode _head = null;
    VaNode _tail = null;
    volatile int _size = 0;

    /**
     * Adds to the tail of the queue.
     * 
     * @param node the node to add
     */
    public void add(VaNode node)
    {
        if (node == null)
            return;
        
        if (_tail == null)
        {
            assert (_head == null) : "VaQueue.enqueue(): unexpectedly found _tail null but head was not null!";
            // Queue is empty, simply add node.
            _head = node;
        }
        else
        {
            // Add node to next of the current tail.
            _tail.next(node);
        }
        
        // update tail to be the node passed in.
        _tail = node;
        _tail.next(null);
        _size++;
    }
    
    /**
     * Removes and returns the head of the queue.
     * 
     * @return the head of the queue
     */
    public VaNode poll()
    {
        if (_head == null)
        {
            return (VaNode)null;
        }
        
        VaNode node = _head;
        _size--;
        
        if(_head.next() == null)
        {
            // queue is empty
            _head = null;
            _tail = null;
            _size = 0;
        }
        else
        {
            // make the next node the new head
            _head = _head.next();
        }
        
        return node;
    }
    
    /**
     * Returns but does not remove the head of the queue.
     * 
     * @return the head of the queue
     */
    public VaNode peek()
    {
        return _head;
    }
    
    /**
     * Removes a node from the queue.
     * 
     * @param node the node to remove
     * 
     * @return true if the node was in the queue, or false if the node wasn't
     */
    public boolean remove(VaNode node)
    {
        if (node == null || _head == null)
            return false;
        
        if (_head == node)
        {
            if (_head.next() != null)
            {
                _head = _head.next();
            }
            else
            {
                _head = null;
                _tail = null;
            }
            _size--;
            return true;
        }
        
        VaNode previousNode = _head;
        VaNode currentNode = _head.next();
        while (currentNode != null)
        {
            if (currentNode == node)
            {
                if (currentNode.next() != null)
                {
                    previousNode.next(currentNode.next());
                }
                else
                {
                    // at the tail.
                    previousNode.next(null);
                    _tail = previousNode;
                }
                _size--;
                return true;
            }
            
            // move to the next node
            previousNode = currentNode;
            currentNode = currentNode.next();
        }
        return false;
    }
    
    /**
     * Returns the size of the queue.
     * 
     * @return the size of the queue
     */
    public int size()
    {
        return _size;
    }
}

File: CSharp/Eta/Src/ValueAdd/Common/VaQueue.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2022-2024 Refinitiv. All rights reserved.         --
 *|-----------------------------------------------------------------------------
 */

using System.Diagnostics;
using System.Runtime.CompilerServices;

namespace LSEG.Eta.ValueAdd.Common
{
    /// <summary>
    /// A FIFO queue.
    /// </summary>
    public class VaQueue
    {
        /// <summary>
        /// The reference to the <see cref="VaNode"/> at the head of the queue.
        /// </summary>
        protected VaNode? _head;

        /// <summary>
        /// The reference to the <see cref="VaNode"/> at the tail of the queue.
        /// </summary>
        protected VaNode? _tail;
        private int _size;

        /// <summary>
        /// Adds to the tail of the queue.
        /// </summary>
        /// <param name="node">The node to add</param>
        [MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.AggressiveInlining)]
        virtual public void Add(VaNode node)
        {
            if (node is null)
                return;

            if (_tail is null)
            {
                Debug.Assert(_head is null, "VaQueue.Add(): unexpectedly found _tail null but head was not null!");
                // Queue is empty, simply add node.
                _head = node;
            }
            else
            {
                // Add node to next of the current tail.
                _tail.Next = node;
            }

            // update tail to be the node passed in.
            _tail = node;
            _tail.Next = null;
            _size++;
        }

        /// <summary>
        /// Polls a <see cref="VaNode"/> from this queue.
        /// </summary>
        /// <returns>Removes the oldest <see cref="VaNode"/></returns>
        [MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.AggressiveInlining)]
        virtual public VaNode? Poll()
        {
            if (_head == null)
            {
                return null;
            }

            VaNode? node = _head;
            _size--;

            if (_head.Next == null)
            {
                // queue is empty
                _head = null;
                _tail = null;
                _size = 0;
            }
            else
            {
                // make the next node the new head
                _head = _head.Next;
            }

            return node;
        }

        /// <summary>
        /// Returns but does not remove the head of the queue.
        /// </summary>
        /// <returns>The head of the queue</returns>
        [MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.AggressiveInlining)]
        virtual public VaNode? Peek()
        {
            return _head;
        }

        /// <summary>
        /// Removes a node from the queue.
        /// </summary>
        /// <param name="node">The node to remove</param>
        /// <returns><c>true</c> if the node was in the queue, or <c>false</c> if the node wasn't</returns>
        [MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.AggressiveInlining)]
        virtual public bool Remove(VaNode node)
        {
            if (node == null || _head == null)
                return false;

            if (_head == node)
            {
                if (_head.Next != null)
                {
                    _head = _head.Next;
                }
                else
                {
                    _head = null;
                    _tail = null;
                }
                _size--;
                return true;
            }

            VaNode? previousNode = _head;
            VaNode? currentNode = _head.Next;
            while (currentNode != null)
            {
                if (currentNode == node)
                {
                    if (currentNode.Next != null)
                    {
                        previousNode.Next = currentNode.Next;
                    }
                    else
                    {
                        // at the tail.
                        previousNode.Next = null;
                        _tail = previousNode;
                    }
                    _size--;
                    return true;
                }

                // move to the next node
                previousNode = currentNode;
                currentNode = currentNode.Next;
            }

            return false;
        }

        /// <summary>
        /// Returns the size of the queue.
        /// </summary>
        /// <returns>The size of the queue</returns>
        [MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.AggressiveInlining)]
        virtual public int Size()
        {
            return _size;
        }

        [MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.AggressiveInlining)]
        internal void VerifyQueue()
        {
            int i = 0;
            Debug.Assert(_size >= 0);
            for (VaNode? node = _head; node != null; node = node.Next)
            {
                ++i;
                if (node.Next == null)
                    Debug.Assert(node == _tail);
            }

            Debug.Assert(i == _size);
        }
    }
}
